Feature/complex secrets and Set-Secret Edit / Add mode

itfranck opened the pull request:
This is #6 with less questions and more code. 
This do include #7 changes too since I couldn't do anything without them (Set-Secret was not working, after all) 

For the context of this PR, 
Complex type = any credential type included with lastpass that is not a note or user/password type. 
(eg: Bank account, credit card, etc...)

Custom type = any custom type defined by the org. and / or user 


This PR goal is to address the following: 
- Get-Secret with complex types (Bank account, credit card) and Custom user defined type currently return only the Notes fields
- Set-Secret to either add or edit existing secret (In main, it always add with no regards for existing secret) 
- Add a **outputType** parameter, that can be set to **Detailed** so normal credentials return as a hashtable containing all the fields instead of a PsCredential (so we get the Notes / Url) fields 
(Maybe this behavior should be the default and instead a outputType with Simple added to return "pscredential" for simple credential and / or maybe there shouldn't be any parameter and everything should be a hashtable ?) 
- In Set-Secret, changed initial use of ConvertFrom-SecureString (which... did not have the -AsPlainText parameter) for the Interop statement, which work with 5.1 (and still work with no issue with 7.1)


Some rough edges. 

- Could not find specified account(s)  error message when using Set-Secret to Add a secret. 
![image](https://user-images.githubusercontent.com/1980296/97106214-9d022100-1696-11eb-8b15-37387df2f1dc.png)
Really not sure why it happens. I never saw an ErrorAction SilentlyContinue being disrespected like that. Maybe it is tied to the `&`. 

- Currently, multiline notes that would have a string that start with something that looks like a key / value pair used by lastpass will vause issue when comes the time to add / edit the secret. 
Take for instance this note 
```
I am a silly note
For helpdesk support, contact 1-333-333-3333
Pin: 9000
```
That last line when sent via input is problematic because it is exactly how key & values are defined. 
There's a Notes parameter but I wasn't successful to pass a multiline string there. 
(It might just be that I tried to escape wrong or that multiline notes using --notes along with (or without) input object are not supported. 

- Complex note types
```
$NoteTypeArgs = @("--note-type=$($Secret.NoteType.ToLower().replace(' ','-'))") 
```
Currently, I used that statement.
Tested with Credit Card but I need to test with other complex types
This is because the `lpass show` return "Credit Card" as note type but it want a specific set of valid type when using the --note-type parameter which seems to lower case / dash as space types (eg: credit-card).

That being said, there's a dozen of type so maybe a map will be required if the translation do not work out of the box for the defined types. 

(Also, any custom types is not supported when adding secret (that's a lpass CLI limitation)

itfranck committed:
Managing Get / Set with support for Complex secret

itfranck committed:
Update README.md

itfranck committed:
Remove forced verbose + old commented code

itfranck committed:
Update SecretManagement.LastPass.Extension.psm1

itfranck committed:
ConvertFrom-SecureString

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -20,13 +20,24 @@ function Invoke-lpass {
 
-    $lpassCommand = if ($null -ne $AdditionalParameters.lpassCommand){$AdditionalParameters.lpassCommand} else {'lpass'}
+    $lpassCommand = if ($null -ne $AdditionalParameters.lpassCommand){$AdditionalParameters.lpassCommand} else {''}
+    $lpassPath = if ($null -ne $AdditionalParameters.lpassPath){"`"$($AdditionalParameters.lpassPath)`""} else {'lpass'}
```
Theses are the changes from #7 since without them, nothing can be tested on the Set-Secret end, they will disappear from the PR once #7 get merged (I will transpose any changes made in the other PR here as needed)

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -50,20 +61,46 @@ function Get-Secret
+        $Output = Get-ComplexSecret -Fields $MyMatches -Note $Note -Raw $Raw
+    }
+    else {
+        $Output = Get-SimpleSecret -Fields $MyMatches -Note $Note
```
That's what make the distinction between 'Detailed' and the default. I preserved the default behavior of returning the Notes as string and regular user / password as psCredential.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
     )
-    if($Secret -is [string]) {
-        $Secret = @{
-            URL = "http://sn"
```
I am 99% sure that the default URL scheme of http://sn is required at all but I'll review this tomorrow just in case.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
+        $Secret = @{Username = $Secret.Username; Password = $Secret.GetNetworkCredential().password}
     }
 
-    if($Secret.Password) {
```
Everything is a hashtable  (mostly) but in the case it is  a String of PSCredential, they are converted so the block below is not needed anymore.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
 
-        $sb.Append("Password: ").AppendLine($pass)
-        $sb.Append("password: ").AppendLine($pass)
-        $sb.Append("sudo_password: ").AppendLine($pass)
```
I really don't understand this. Why 2 times password and why sudo_password.
From my tests, these were not fields from the lpass CLI when getting default credentials so this has not been replicated.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
-    }
+    
+    if ($Secret -is [hashtable]){
+        $SpecialKeys = @('Language', 'NoteType', 'Notes')
```
Custom types have a Language hidden field.
Complex / Custom types have a NoteType field
Notes fields, being the only multiline field, is a special one and dealt with accordingly.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
+        foreach ($k in $Keys) {
+            if ($Secret.$k -is [securestring]) {
+                $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.$k)
+                $Secret.$k = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
```
The Windows 5.1 way (also work in 7+)

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
     try {
-        $sb.ToString() | Invoke-lpass 'add', $Name, '--non-interactive'
-    } catch {
+        $res = Invoke-lpass 'show', '--sync=now', '--name', $Name -ErrorAction SilentlyContinue
```
This one I really do not understand...  When adding a secret and doing a "Show", it return the error  Could not find specified account(s). This would be expected if I did not put -ErrorAction SilentlyContinue.

itfranck replied:
Redirecting the error stream through `someExe --argument 2>$null`
https://stackoverflow.com/questions/35940593/suppressing-powershell-errors-when-calling-external-executables

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
+          $SecretExists = $null -ne $res 
+
+        if ($SecretExists) {
+            Write-Verbose "Editing secret" 
```
Verbose won't work for now.
https://github.com/PowerShell/SecretManagement/issues/66

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
+                    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.NoteType)
+                    $Secret.NoteType = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+                }
+                 $NoteTypeArgs = @("--note-type=$($Secret.NoteType.ToLower().replace(' ','-'))") 
```
This, I need to test against the dozen of existing complex types to see if it hold. Otherwise, a hashtable to link what we get from the GET to what the SET expect will be required.

itfranck replied:
I added a key map to deal with this since there was not a way to convert the type between display / internal name easily.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -79,46 +116,66 @@ function Set-Secret
+                }
+                 $NoteTypeArgs = @("--note-type=$($Secret.NoteType.ToLower().replace(' ','-'))") 
+            }
+            $sb.ToString() | Invoke-lpass 'add', $Name, '--non-interactive', $NoteTypeArgs
```
I've attempted to use the Notes arguments here to resolve the issue of semi-column from the input... I was not successful but if it work, it would resolve the second moot point.

itfranck replied:
Current brain think old brain needed a break... Not so complicated to address.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -159,11 +216,17 @@ function Get-SecretInfo
+                    [SecretType]::PSCredential
+                }
+                else {
+                    [SecretType]::Unknown
```
I can't know if it's a note or a custom record, hence "unknown" here.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -183,3 +246,67 @@ function Test-SecretVault
+        Write-Warning -Message @"
+The record contains multiple fields with the same name.
+$Dupesstr
+A Raw parameter will be added to the returned object with the raw object. Only the first field found with a duplicated name will be included in the parsed object.
```
Not sure it's the best way to deal with duplicates but since we cannot use a hashtable, I thought it would be better to return somewhat the secret with a raw field rather than nothing at all and or the unparsed string.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -183,3 +246,67 @@ function Test-SecretVault
+    $Output = @{}
+    if (![String]::IsNullOrEmpty($Note)) { 
+        $Output.Notes = $Note
+        $Fields = $Fields | Select-Object -SkipLast 1
```
Notes is always last. Even if you create a custom type, add the Notes field first and add something afterward, it get slided down.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -183,3 +246,67 @@ function Test-SecretVault
     return (Get-Command lpass -ErrorAction SilentlyContinue) -and ((lpass status) -match "Logged in as .*")
 }
+
+function Get-SimpleSecret {
```
I think I didn't deal with secure strings here so I'll review that too.

itfranck replied:
Nevermind... Secure string need to be dealt with in the set, not the Get.

TylerLeonhardt commented:
can you rebase/merge from master so I can see the actual change here?

I'm really excited for this feature!

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -61,20 +61,46 @@ function Get-Secret
+    $Raw = ($res | Select-Object -Skip 1) -join "`n"
+
+    # Custom type cannot have the same case-sensitive name but Username / uSERname work.
+    # That's why we don't parse it directly to a hashtable.
```
Maybe use this solution for a case-sensitive hashtable?

https://stackoverflow.com/a/24054257

If that doesn't work, maybe just use a Dictionary:

```pwsh
$myDict = [System.Collections.Generic.Dictionary[string,string]]::new()
$myDict['foo'] = "bar"
$myDict['FOo'] = "bar"
```

itfranck replied:
Ohhh indeed. I'll test that.

itfranck committed:
Using sensitive hashtable where required.

itfranck committed:
Update README.md

itfranck committed:
Redirect error stream from external command call

Discard lpass error stream

itfranck committed:
Revert error stream redirection

itfranck committed:
Deal with native command error

itfranck committed:
Update SecretManagement.LastPass.Extension.psm1

itfranck committed:
Added secret map.

This is to address the discrepancies between Get / Set operation in lPass.

itfranck committed:
bugfixes

itfranck committed:
Update SecretManagement.LastPass.Extension.psm1

itfranck committed:
Need to use the first case-sensitive match, not the last.

itfranck committed:
Added back support for simple notes with Set-Secret

Providing the URL parameter with a value of http://sn and only a Notes field create the record as a secure note instead of a credential.

itfranck committed:
Fix redirect to null + explicit sync

itfranck commented:
@TylerLeonhardt Every issues / concerns have been fixed. 
Feel free to review when you have some time. 
Let me know if there's anything you want me to expand further on.

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -26,17 +53,17 @@ function Invoke-lpass {
-            $InputObject | & "$lpassCommand" $lpassPath @Arguments
+            return $InputObject | & "$lpassCommand" $lpassPath @Arguments 
         }
         else {
```
you can also get rid of this `else {` since the `if` returns.

```pwsh
if ($InputObject) {
    # ...
    return ...
}

return ...
```

itfranck replied:
Indeed, there was no technical reason. I won't usually put an else if I return but since both side were one liner and already part of a if in the parent level, it just felt easier to the eye to me. I removed them.

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -26,17 +53,17 @@ function Invoke-lpass {
-            $InputObject | & $lpassPath @Arguments
+            return  $InputObject | & $lpassPath @Arguments 
         }
         else {
```
likewise

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -61,20 +88,55 @@ function Get-Secret
+    try {
+        $res = Invoke-lpass 'show', '--name', $Name, '--all'
+        if ($null -eq $res) {
+            return 
```
should this just return? What will the user see?

itfranck replied:
By returning nothing, we leave to the SecretManagement module the error throw.
The following error will be displayed to the user: Get-Secret : The secret $Name was not found.

I added a comment in the code to make this fact clear.

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -61,20 +88,55 @@ function Get-Secret
+            return 
+        }
+        if ($res[0] -eq 'Multiple matches found.') {
+            Write-Warning "Multiple matches found with the name $Name. `nThe first matching result will be returned."
```
Couldn't you run the remaining code in this function on each result instead of grabbing just 1?

itfranck replied:
I believe Get-Secret can only return one secret. 
I don't have a definitive proof but based on the wildcard discussion on the main project, I assumed the reason they didn't support wildcards was exactly that. 

Get-Secret is meant to return only 1 error.
That being said, I'll validate my assumptions.

itfranck replied:
Yup. First secret only. 

 This cmdlet finds and returns the first secret that matches the provided name. If a vault name is specified, then only that vault will be searched. Otherwise, all vaults are searched and the first found result is returned...

itfranck replied:
@TylerLeonhardt FYI

Get-Secret only return the first secret if multiples are found. 
I attempted to return a list of secrets by changing my `return $output` for a list of 2 secrets and I was able to confirm it first hand.

TylerLeonhardt replied:
Interesting... I wonder if that's feedback on SecretManagement. In theory a vault could have multiple secrets of the same name...

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -61,20 +88,55 @@ function Get-Secret
+        return 
+    }
+    
+    $Raw = ($res | Select-Object -Skip 1) -join "`n"
```
Can you leave a comment as to why you are skipping 1.

itfranck replied:
First line is the secret name and ID so we don't want / need  it. I added a comment in the code to this effect..

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -61,20 +88,55 @@ function Get-Secret
+        })
+
+    # Notes is always the last item. This is also the only field that can be multiline.
+    $HasNote = $MyMatches.key -ccontains 'Notes' 
```
if Notes is always the last item, could you do this?:

```pwsh
$HasNote = $MyMatches[$MyMatches.Count - 1].key -eq 'Notes'
```

Then you don't have to iterate through the keys.

You could probably use this for getting the start index as well on line 127.

itfranck replied:
@TylerLeonhardt  Consider the following. 

- Notes field can exist only once.
- Notes field can be multiline
- Notes field can contains something like : 

Notes:
```
Here are the notes.
Nothing so tricky
Notes: This is tricky
Pin: 1234
```

The `$MyMatches` regex pattern `(?<key>.*?)\: (?<value>.*?)(?:\n|$)` will pick up the legitimate fields but also everything in the Notes field that match the pattern, `Notes: This is tricky` and `Pin: 1234` 

`Notes` is always the last valid field returned by lpass CLI.
The `$MyMatches` pattern will also parse the legitimate Notes field into key  / values if they match the pattern. 

Thus, while the `Notes` field is the last field from lpass CLI, it won't be the last in `$MyMatches` if the notes value contains things that were parsed by the pattern (eg: Second Notes and Pin as per above example). For that reason, the first encounter with the **Notes** field in `$MyMatches` is the one we are looking for. 

Anything afterward is "fake" and need to be discarded.

This is why I discard any matches after Notes, since they are just things that match the pattern but actually belong to the same last multiline Notes item.
```
 $MyMatches = $MyMatches.Where({$_.ValueIndex -lt $start})
```

I added the following comment in the code to make this hopefully clearer.

```
# Notes is always the last item (lpass wise). This is also the only field that can be multiline.
# Any matching items after the first "Notes" key need to be discarded as it is not an item,
# but rather just part of the notes.
```

TylerLeonhardt replied:
That makes perfect sense! Thanks!

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -90,46 +152,72 @@ function Set-Secret
-    }
+    
+    if ($Secret -is [hashtable]){
+        $SpecialKeys = @('Language', 'NoteType', 'Notes')
```
This can be pulled up the top of the file as a "constant"

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -90,46 +152,72 @@ function Set-Secret
+                $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.$k)
+                $Secret.$k = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+            }
+            [Void]($sb.AppendLine("$($k): $($Secret.$k)"))
```
```suggestion
            $sb.AppendLine("$($k): $($Secret.$k)") | Out-Null
```

Out-Null is fine here. It's the fastest in PS7.

itfranck replied:
Nice to know.

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -90,46 +152,72 @@ function Set-Secret
+                $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.Notes)
+                $Secret.Notes = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+            }
+            [Void]($sb.AppendLine("Notes: `n$($Secret.Notes)"))
```
```suggestion
            $sb.AppendLine("Notes: `n$($Secret.Notes)") | Out-Null
```

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -90,46 +152,72 @@ function Set-Secret
+                    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.NoteType)
+                    $Secret.NoteType = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
+                }
+                $NoteTypeArgs = @("--note-type=$($Secret.NoteType)")
```
```suggestion
                $NoteTypeArgs += "--note-type=$($Secret.NoteType)"
```
If you're allocating a new array anyway, might as well do it this way.

itfranck replied:
It's a little bit cleaner but it feels like I'm encouraging anybody that would want to expand this array later on for any reason to reuse `+=`. 

That's why I never used `+=` for array, even once.

itfranck replied:
Done anyway :)

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -170,11 +258,17 @@ function Get-SecretInfo
-            $type = if ($Matches[3]) {
-                [SecretType]::PSCredential
+            if ($AdditionalParameters.outputType -eq 'Detailed') {
+                $type = [SecretType]::Hashtable
             } else {
-                [SecretType]::SecureString
+                $type = if ($Matches[3]) {
+                    [SecretType]::PSCredential
+                }
+                else {
+                    [SecretType]::Unknown
+                }
             }
```
```pwsh
$type = if ($AdditionalParameters.outputType -eq 'Detailed') {
    [SecretType]::Hashtable
} elseif ($Matches[3]) {
    [SecretType]::PSCredential
} else {
    [SecretType]::Unknown
}
```

a tad cleaner IMO

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -170,11 +258,17 @@ function Get-SecretInfo
+                    [SecretType]::Unknown
+                }
             }
+           
```
nit: looks like there's an extra newline here.

itfranck replied:
👍

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -195,3 +289,59 @@ function Test-SecretVault
+        $Output.NoteType = $DefaultNoteTypeMap.Item($Output.NoteType)
+    }
+    return $Output
+}
```
```suggestion
}

```
nit: newline at EOF

itfranck replied:
TIL:  I mostly work on Windows operating system. 
Good to know and done.

TylerLeonhardt replied:
I just don't like seeing the icon on GitHub when there isn't a newline :)

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -195,3 +289,59 @@ function Test-SecretVault
+        $Note
+    )
+    
+    $IsCredentials = $Fields.where( { $_.key -in @('Username', 'Password') }, 'first').count -eq 1
+    if ($IsCredentials) {
+        $username = $Fields.Where( { $_.key -eq 'Username' }, 'first') | Select-Object -ExpandProperty value
+        if ($null -eq $username) { $username = '' }
+        $password = $Fields.Where( { $_.key -eq 'Password' }, 'first') | Select-Object -ExpandProperty value
+        if ($null -eq $password) { $password = '' }
+        if ("" -ne $password) { $password = $password | ConvertTo-SecureString -AsPlainText -Force }
+        $output = [System.Management.Automation.PSCredential]::new($username, $password)
+    }
+    else {
+        if ($null -eq $Note) { $output = "" } else { $output = $Note }
+    }
+    return $output
```
another possible implementation I think:

```suggestion
    $username = $Fields.Where( { $_.key -eq 'Username' }, 'first') | Select-Object -ExpandProperty value
    $password = $Fields.Where( { $_.key -eq 'Password' }, 'first') | Select-Object -ExpandProperty value
    if ($username -or $password) {
        if ($null -eq $username) { $username = '' }
        if ($null -eq $password) { $password = '' }
        if ("" -ne $password) { $password = $password | ConvertTo-SecureString -AsPlainText -Force }
        return [System.Management.Automation.PSCredential]::new($username, $password)
    }

    if ($null -ne $Note) {
        return $Note
    }

    return ''
```

itfranck replied:
I like it. 
It's cleaner.
(Note: I didn't notice the "Commit suggestion" button until after committing manually. I'll know that for next time)

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -195,3 +289,59 @@ function Test-SecretVault
+    )
+    $Dupes = ($Fields | Group-Object key).Where( { $_.Count -gt 1 })
+    
+    if ($Dupes.count -gt 0) {
```
```suggestion
    if ($Dupes.Count -gt 0) {
```
nit

TylerLeonhardt reviewed (commented):
This is looking really good! Left some comments on cleaning up the code and some questions.

Really excited for this!

itfranck committed:
Requested changes.

itfranck committed:
Indent

itfranck committed:
Requested / suggested changes.

TylerLeonhardt commented:
Not sure if the request for a review was a "I addressed everything" but it looks like there are a couple more comments. 1 important one that could help with performance, and few nits.

itfranck commented:
Ohhhhhh..... There's was a Show hidden conversation with more comments in it !!!!
Looking at the remaining ones now.

itfranck committed:
Update SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

itfranck committed:
Update SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

itfranck committed:
Update SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

itfranck committed:
Moved $SpecialKeys at the top of file.

itfranck committed:
Comments regarding Regex Notes field

itfranck commented:
@TylerLeonhardt Yup, that what it meant.
However, I completely missed the conversations hidden behind "Show more conversations" button. 
I have now addressed the remaining changes 

There's one conversation still open.
No code changes were made but explanation was provided and code comments were added.

TylerLeonhardt reviewed (approved):
LGTM this is awesome! I'll get started on a release today.
